### PR TITLE
gmp: update to 6.2.0 + add components

### LINKED ISSUE #996: [request] gmp/6.2.0
### Package Details
  * Package Name/Version: **gmp/6.2.0**
  * Changelog: **https://gmplib.org**

The above mentioned version is newly released by the upstream project and not yet available as a recipe. Please add this version.

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **gmp/6.2.0**

Fixes #996

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by madebr** [comment]
@danimtb 
Am I using components wrong?
```
ERROR: 
	ConanException: gmp/6.1.2 package_info(): self.cpp_info.components cannot be used with self.cpp_info global values at the same time
```

**Comment by danimtb** [comment]
@madebr if you use components, you are not able to model things in the global scope. That means for example, that you cannot use `self.cpp_info.libs` and components at the same time

**Comment by madebr** [comment]
@danimtb 
Thanks! I misunderstood global scope, I guess.

**Comment by madebr** [comment]
Wow, bitten by my own hook :sweat_smile:

**Comment by madebr** [comment]
@danimtb 
There is an error related to components in test_package.
See https://github.com/conan-io/conan-center-index/pull/1541#issuecomment-624707269

**Comment by danimtb** [comment]
@madebr this is clearly a bug in the JSON serializer for components. I have already proposed a fix. Thanks a lot for reporting!

**Comment by madebr** [comment]
@danimtb @jgsogo @SSE4 @uilianries 
Please review!

File: recipes/gmp/all/conanfile.py [hunk 1]
```diff
@@ -69,19 +69,19 @@ def package(self):
...
+        os.unlink(os.path.join(self.package_folder, "lib", "libgmp.la"))
+        if self.options.enable_cxx:
+            os.unlink(os.path.join(self.package_folder, "lib", "libgmpxx.la"))
 
     def package_id(self):
         del self.info.options.run_checks  # run_checks doesn't affect package's ID
 
     def package_info(self):
+        self.cpp_info.components["libgmp"].libs = ["gmp"]
+        self.cpp_info.components["libgmp"].names["pkg_config"] = ["gmp"]
```
Comment (jgsogo) [inline recipes/gmp/all/conanfile.py:84]: @danimtb how does this work? `names` for `pkg_config` at a component level

**Comment by madebr** [comment]
@jgsogo 
I just saw that I (accidentally) removed the `GMP` name.
Googling for `cmake find gmp` only gives me `FindGMP.cmake` results.

So I'll add a patch that re-instates this.
The `pkg_config` names are the official ones.

**Comment by jgsogo** [comment]
> re your question. conan forbade me using the `gmp` name as a component.
> That's why I opted for `libgmp`.
> I would have been happy with `gmp`.

Yes, that's the reason why I was wondering if this package should have been `libgmp` instead of `gmp`, so you can use `gmp` as a component. 🤔

File: recipes/gmp/all/conanfile.py (hunk 1 above)
Comment (danimtb) [reply recipes/gmp/all/conanfile.py:85]: yes, the syntax is allowed. However, the `pkg_config` generator still does not handle components. Do you know what are the equivalent to components in pkg-config @madebr?

File: recipes/gmp/all/conanfile.py [hunk 2]
```diff
@@ -69,19 +69,25 @@ def package(self):
...
-            self.cpp_info.libs = ["gmpxx"]
-        self.cpp_info.libs.append("gmp")
-        self.cpp_info.names["cmake_find_package"] = "GMP"
-        self.cpp_info.names["cmake_find_package_multi"] = "GMP"
+            super_component = "gmpxx"
+            self.cpp_info.components["gmpxx"].libs = ["gmpxx"]
+            self.cpp_info.components["gmpxx"].requires = ["libgmp"]
+
+        # Make sure the GMP cmake_find_package generator knows about both libgmp and gmpxx
+        self.cpp_info.components[super_component].names["cmake_find_package"] = "GMP"
```
Comment (danimtb) [inline recipes/gmp/all/conanfile.py:92]: So with this "super_component", this means that the ``gmp::GMP`` target will be different depending on the ``enable_cxx`` option? That looks a bit weird to me, but I have no idea how it is made so...

File: recipes/gmp/all/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/gmp/all/conanfile.py:85]: pkg-config pc files and targets have a 1:1 mapping. Each file represents one target (and its dependencies).
See https://people.freedesktop.org/~dbn/pkg-config-guide.html for an example.

In the case of gmp, 2 pc files: `gmp.pc` and `gmpxx.pc` should be created.

File: recipes/gmp/all/conanfile.py (hunk 2 above)
Comment (madebr) [reply recipes/gmp/all/conanfile.py:92]: This indeed seems a bit weird.
I'll change it to `GMP` for `gmp` and `GMPXX` for `gmpxx`.

I don't think people usually the `gmp` library when searching for `gmp`.

File: recipes/gmp/all/conanfile.py [hunk 3]
```diff
@@ -80,14 +80,13 @@ def package_id(self):
...
         if self.options.enable_cxx:
-            super_component = "gmpxx"
             self.cpp_info.components["gmpxx"].libs = ["gmpxx"]
             self.cpp_info.components["gmpxx"].requires = ["libgmp"]
-
-        # Make sure the GMP cmake_find_package generator knows about both libgmp and gmpxx
-        self.cpp_info.components[super_component].names["cmake_find_package"] = "GMP"
-        self.cpp_info.components[super_component].names["cmake_find_package_multi"] = "GMP"
+            self.cpp_info.components["gmpxx"].names["cmake_find_package"] = "GMP"
+            self.cpp_info.components["gmpxx"].names["cmake_find_package_multi"] = "GMP"
```
Comment (madebr) [inline recipes/gmp/all/conanfile.py:91]:
```suggestion
            self.cpp_info.components["gmpxx"].names["cmake_find_package"] = "GMPXX"
            self.cpp_info.components["gmpxx"].names["cmake_find_package_multi"] = "GMPXX"
```

**Pull request merged**